registry/distribution/repository.go

```diff
@@ -0,0 +1,674 @@
+func (s *blobStore) Push(ctx context.Context, expected ocispec.Descriptor, content io.Reader) error {
+	// start an upload
+	url := fmt.Sprintf("%s/blobs/uploads/", s.repo.endpoint())
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, nil)
+	if err != nil {
+		return err
+	}
+
+	resp, err := s.repo.Client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusAccepted {
+		defer resp.Body.Close()
+		return parseErrorResponse(resp)
+	}
+	resp.Body.Close()
+
+	// monolithic upload
+	location, err := resp.Location()
+	if err != nil {
+		return err
+	}
+	url = location.String()
+	req, err = http.NewRequestWithContext(ctx, http.MethodPut, url, content)
+	if err != nil {
+		return err
+	}
+	if req.GetBody != nil && req.ContentLength != expected.Size {
+		// short circuit a size mismatch for built-in types.
+		return fmt.Errorf("mismatch content length %d: expect %d", req.ContentLength, expected.Size)
+	}
+	req.ContentLength = expected.Size
+	// the expected media type is ignored as in the API doc.
+	req.Header.Set("Content-Type", "application/octet-stream")
+	q := req.URL.Query()
+	q.Set("digest", expected.Digest.String())
+	req.URL.RawQuery = q.Encode()
+
+	resp, err = s.repo.Client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusCreated {
+		return parseErrorResponse(resp)
+	}
+	return nil
+}
+
+// Exists returns true if the described content exists.
+func (s *blobStore) Exists(ctx context.Context, target ocispec.Descriptor) (bool, error) {
+	_, err := s.Resolve(ctx, target.Digest.String())
+	if err == nil {
+		return true, nil
+	}
+	if errors.Is(err, errdef.ErrNotFound) {
+		return false, nil
+	}
+	return false, err
+}
+
+// Delete removes the content identified by the descriptor.
+func (s *blobStore) Delete(ctx context.Context, target ocispec.Descriptor) error {
+	return s.repo.delete(ctx, "blobs", target)
+}
+
+// Resolve resolves a reference to a descriptor.
+func (s *blobStore) Resolve(ctx context.Context, reference string) (ocispec.Descriptor, error) {
+	ref, err := s.repo.parseReference(reference)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	refDigest, err := ref.Digest()
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	url := fmt.Sprintf("%s/blobs/%s", s.repo.endpoint(), refDigest)
+	req, err := http.NewRequestWithContext(ctx, http.MethodHead, url, nil)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+
+	resp, err := s.repo.Client.Do(req)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
+		// no-op
+	case http.StatusNotFound:
+		return ocispec.Descriptor{}, fmt.Errorf("%s: %w", ref, errdef.ErrNotFound)
+	default:
+		return ocispec.Descriptor{}, parseErrorResponse(resp)
+	}
+	mediaType, _, _ := mime.ParseMediaType(resp.Header.Get("Content-Type"))
+	if mediaType == "" {
```

Comment:
Will the response mediaType ever be empty?

Reply:
It depends on the server implementation. Not all registries implement correctly.